src/Our.Umbraco.Ditto/Proxy/ProxyFactory.cs

```diff
@@ -0,0 +1,287 @@
+﻿namespace Our.Umbraco.Ditto
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using System.Reflection.Emit;
+    using System.Threading;
+
+    /// <summary>
+    /// The proxy factory for creating instances of proxy classes.
+    /// </summary>
+    public class ProxyFactory
+    {
+        /// <summary>
+        /// Ensures that proxy creation is atomic.
+        /// </summary>
+        private static readonly ReaderWriterLockSlim Locker = new ReaderWriterLockSlim();
+
+        /// <summary>
+        /// The proxy cache for storing proxy types.
+        /// </summary>
+        private static readonly ConcurrentDictionary<ProxyCacheEntry, Type> ProxyCache = new ConcurrentDictionary<ProxyCacheEntry, Type>();
+
+        /// <summary>
+        /// Creates an instance of the proxy class for the given <see cref="Type"/>.
+        /// </summary>
+        /// <param name="baseType">
+        /// The base <see cref="Type"/> to proxy.
+        /// </param>
+        /// <param name="interceptor">
+        /// The <see cref="IInterceptor"/> to intercept properties with.
+        /// </param>
+        /// <param name="excludedProperties">
+        /// The <see cref="IEnumerable{PropertyInfo}"/> of properties to exclude.
+        /// </param>
+        /// <param name="constructorArguments">
+        /// The constructor arguments to pass to any proxy types.
+        /// </param>
+        /// <returns>
+        /// The proxy <see cref="Type"/> instance.
+        /// </returns>
+        public object CreateProxy(Type baseType, IInterceptor interceptor, IEnumerable<PropertyInfo> excludedProperties, params object[] constructorArguments)
+        {
+            Type proxyType = this.CreateProxyType(baseType, excludedProperties);
+
+            object result = Activator.CreateInstance(proxyType, constructorArguments);
```

Comment:
That was the issue. `TypeInitializationExtensions.GetInstance` can't accept params, it can't determine the type properly. I'll probably create two separate method signatures since we know the two options.